File: day04/main.go
package main

import (
	"os"
	"regexp"
	"strconv"
	"strings"
)

type (
	Context struct {
		CardRegex *regexp.Regexp
	}

	Match struct {
		Number   int
		Player   []int
		Expected map[int]struct{}
	}

	Stringable interface {
		String() string
	}
)

const (
	CARD_REGEX = "Card (\\d+): "
)

func NewContext() *Context {
	cardRegex, err := regexp.Compile(CARD_REGEX)
	idgaf(err)
	return &Context{
		CardRegex: cardRegex,
	}
}

func NewMatch(number int, player []int, expected map[int]struct{}) *Match {
	return &Match{
		Number:   number,
		Player:   player,
		Expected: expected,
	}
}

func (m *Match) String() string {
	return strconv.Itoa(m.Number) + ": " + ToPrintable(m.Player) + " | " + ToPrintable(m.Expected)
}

func idgaf(err error) {
	if err != nil {
		panic(err)
	}
}

func ToPrintable[T Stringable](slice []T) string {
	s := "["
	for _, v := range slice {
		s += v.String() + ", "
	}
	return s + "]"
}

func part1(ctx *Context, lines []string) {
	matches := parse1(ctx, lines)
	println(ToPrintable(matches))
}

func parseLine1(ctx *Context, line string) *Match {
	gameNumber, err := strconv.Atoi(ctx.CardRegex.FindStringSubmatch(line)[1])
	idgaf(err)
	plays := ctx.CardRegex.ReplaceAllLiteralString(line, "")

	tmp := strings.Split(plays, "| ")
	rawPlayer := strings.Split(tmp[0], " ")
	rawExpected := strings.Split(tmp[1], " ")

	player := []int{}
	for _, p := range rawPlayer {
		card, err := strconv.Atoi(strings.TrimSpace(p))
		if err == nil {
			player = append(player, card)
		}
	}

	expected := map[int]struct{}{}
	for _, e := range rawExpected {
		card, err := strconv.Atoi(strings.TrimSpace(e))
		if err == nil {
			expected[card] = struct{}{}
		}
	}

	return NewMatch(gameNumber, player, expected)
}

func parse1(ctx *Context, lines []string) []*Match {
	matches := []*Match{}
	for _, line := range lines {
		matches = append(matches, parseLine1(ctx, line))
	}
	return matches
}

func main() {
	ctx := NewContext()
	fileName := "data/test.txt"
	file, err := os.ReadFile(fileName)
	idgaf(err)
	contents := string(file)
	lines := strings.Split(contents, "\n")

	part1(ctx, lines)
}

File: day04/main.py
import re
from collections import defaultdict
from dataclasses import dataclass

CARD_REGEX = r"Card\s*(\d+): "


@dataclass
class Match:
    number: int
    player: set[int]
    expected: set[int]


class Part2:
    def __init__(self, line: list[str]):
        self.cards = defaultdict(int)
        self.matches = [parse_line1(line) for line in lines]

    def play(self, idx: int) -> int:
        if idx >= len(self.matches):
            return 0
        self.cards[idx] += 1
        current_match = self.matches[idx]
        counter = sum(
            1 for card in current_match.player if card in current_match.expected
        )
        i = idx + 1
        while counter > 0:
            self.play(i)
            i += 1
            counter -= 1

    def run(self):
        for i in range(len(self.matches)):
            self.play(i)
        return sum(v for v in self.cards.values())


def parse_line1(line: str):
    game_number = int(re.match(CARD_REGEX, line).group(1))
    _, plays = line.split(": ")
    e, p = plays.split("| ")
    player = {int(x.strip()) for x in p.split(" ") if x.strip()}
    expected = {int(x.strip()) for x in e.split(" ") if x.strip()}
    return Match(game_number, player, expected)


def play1(match: Match) -> int:
    points = 0
    multiplier = 2
    for card in match.player:
        if card in match.expected:
            points = 1 if points == 0 else points * multiplier
    return points


def part1(lines: list[str]) -> int:
    matches = [parse_line1(line) for line in lines]
    return sum(play1(match) for match in matches)


filename = "data/input.txt"
with open(filename, "r") as fin:
    lines = fin.readlines()

print(part1(lines))
part2 = Part2(lines)
print(part2.run())
